### PR TITLE
Add Dsytrd and tests

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

File: native/dsytrd.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	"github.com/gonum/blas/blas64"
+)
+
+// Dsytrd reduces a symmetric n×n matrix A to symmetric tridiagonal form by an
+// orthogonal similarity transformation.
+//  Q^T * A * Q = T
+// where Q is an orthonormal matrix and T is symmetric and tridiagonal.
+//
+// On entry, a contains the elements of the input matrix in the triangle specified
+// by uplo. On exit, the diagonal and sub/superdiagonal are overwritten by the
```
Comment (kortschak) [inline native/dsytrd.go]: s/superdiagonal/super-diagonal/

File: native/dsytrd.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+				}
+			}
+		} else {
+			nx = n
+		}
+	} else {
+		nb = 1
+	}
+	ldwork = nb
+	if upper {
```
Comment (kortschak) [inline native/dsytrd.go]: Blank line above this.

File: native/dsytrd.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+					nx = n
+				}
+			}
+		} else {
+			nx = n
+		}
+	} else {
+		nb = 1
+	}
+	ldwork = nb
```
Comment (kortschak) [inline native/dsytrd.go]: Why is this here?

Comment (btracey) [inline native/dsytrd.go]: In Fortran, ldwork = n. Here it is nb. nb can be modified in the block above, so it's reset here.

File: native/dsytrd.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+		var i int
+		// Reduce the lower triangle of A.
+		for i = 0; i < n-nx; i -= nb {
+			// Reduce columns 0:i+nb to tridiagonal form and form the matrix W
+			// which is needed to update the unreduced part of the matrix.
+			impl.Dlatrd(uplo, n-i, nb, a[i*lda+i:], lda, e[i:], tau[i:], work, ldwork)
+
+			// Update the unreduced submatrix A[i+ib:n, i+ib:n], using an update
+			// of the form A -= V*W^T - W*V^T.
+			bi.Dsyr2k(uplo, blas.NoTrans, n-i-nb, nb, -1, a[(i+nb)*lda+i:], lda, work[nb+i:], ldwork, 1, a[(i+nb)*lda+i+nb:], lda)
```
Comment (kortschak) [inline native/dsytrd.go]: Split this.

File: native/dsytrd.go (hunk 1 above)
Comment (btracey) [inline native/dsytrd.go]: Done.

File: native/dsytrd.go (hunk 2 above)
Comment (btracey) [inline native/dsytrd.go]: Done.

File: native/dsytrd.go (hunk 4 above)
Comment (btracey) [inline native/dsytrd.go]: Done. Sorry.

File: native/dsytrd.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+// Dsytrd reduces a symmetric n×n matrix A to symmetric tridiagonal form by an
+// orthogonal similarity transformation.
```
Comment (vladimir-ch) [inline native/dsytrd.go]: Remove dot at the end.

Comment (btracey) [inline native/dsytrd.go]: Done.

**Comment by btracey** [comment]
PTAL

**Comment by kortschak** [comment]
LGTM but wait for @vladimir-ch.

File: native/dsytrd.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+//  [ d            ]
+//  [ e  d         ]
+//  [v1  e  d      ]
+//  [v1 v2  e  d   ]
+//  [v1 v2  v3 e  d]
+//
+// d must have length n, and e and tau must have length n-1. Dsytrd will panic if
+// these conditions are not met.
+//
+// Work is temporary storage, and lwork specifies the usable memory length. At minimum,
```
Comment (vladimir-ch) [inline native/dsytrd.go]: Lowercase w in `Work`?

File: native/dsytrd.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+		// block is always handled by unblocked code.
+		opts := "L"
+		if upper {
+			opts = "U"
+		}
+		nx = max(nb, impl.Ilaenv(3, "DSYTRD", opts, n, -1, -1, -1))
+		if nx < n {
+			// Determine if workspace is large enough for blocked code.
+			ldwork = nb
+			iws := n * ldwork
```
Comment (vladimir-ch) [inline native/dsytrd.go]: This variable is created for the sake of similarity with Fortran?

File: native/dsytrd.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		// unblocked method.
+		var i int
+		kk := n - ((n-nx+nb-1)/nb)*nb
+		for i = n - nb; i >= kk; i -= nb {
+			// Reduce columns i:i+nb to tridiagonal form and form the matrix W
+			// which is needed to update the unreduced part of the matrix.
+			impl.Dlatrd(uplo, i+nb, nb, a, lda, e, tau, work, ldwork)
+
+			// Update the unreduced submatrix A[0:i-1,0:i-1], using an update
+			// of the form A -= V*W^T - W*V^T.
```
Comment (vladimir-ch) [inline native/dsytrd.go]: Written like this there should be a `+` at the rhs.

File: native/dsytrd.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+				d[j] = a[j*lda+j]
+			}
+		}
+		// Use unblocked code to reduce the last or only block
+		// check that i == kk.
+		impl.Dsytd2(uplo, kk, a, lda, d, e, tau)
+	} else {
+		var i int
+		// Reduce the lower triangle of A.
+		for i = 0; i < n-nx; i -= nb {
```
Comment (vladimir-ch) [inline native/dsytrd.go]: Shouldn't be there `i += nb`?

File: native/dsytrd.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+			// Reduce columns 0:i+nb to tridiagonal form and form the matrix W
+			// which is needed to update the unreduced part of the matrix.
+			impl.Dlatrd(uplo, n-i, nb, a[i*lda+i:], lda, e[i:], tau[i:], work, ldwork)
+
+			// Update the unreduced submatrix A[i+ib:n, i+ib:n], using an update
+			// of the form A -= V*W^T - W*V^T.
```
Comment (vladimir-ch) [inline native/dsytrd.go]: `+` also here or write `A = A - ...`.

**Comment by vladimir-ch** [comment]
LGTM after comments addressed.

File: native/dsytrd.go (hunk 6 above)
Comment (btracey) [inline native/dsytrd.go]: Done.

File: native/dsytrd.go (hunk 7 above)
Comment (btracey) [inline native/dsytrd.go]: Yea, for now.

File: native/dsytrd.go (hunk 8 above)
Comment (btracey) [inline native/dsytrd.go]: Oops, thanks.

File: native/dsytrd.go (hunk 10 above)
Comment (btracey) [inline native/dsytrd.go]: Done.

File: native/dsytrd.go (hunk 9 above)
Comment (btracey) [inline native/dsytrd.go]: Oops. Thanks!

**Comment by kortschak** [comment]
I just noticed in other work that this does this:

```
--- FAIL: TestDsytrd (0.01s)
	dsytrd.go:68: A mismatch: upper = true, n = 50
	dsytrd.go:71: D mismatch: upper = true, n = 50
	dsytrd.go:74: E mismatch: upper = true, n = 50
	dsytrd.go:77: Tau mismatch: upper = true, n = 50
panic: runtime error: slice bounds out of range [recovered]
	panic: runtime error: slice bounds out of range

goroutine 57331 [running]:
testing.tRunner.func1(0xc8203a41b0)
	/home/daniel/Development/go/src/testing/testing.go:467 +0x192
github.com/gonum/blas/native.Implementation.Dsyr2k(0x79, 0x6f, 0x44, 0x20, 0xbff0000000000000, 0xc82030c220, 0x26cc, 0x26cc, 0x64, 0xc820320000, ...)
	/home/daniel/Development/src/github.com/gonum/blas/native/level3double.go:562 +0x998
github.com/gonum/blas/native.(*Implementation).Dsyr2k(0x85f6f8, 0x79, 0x6f, 0x44, 0x20, 0xbff0000000000000, 0xc82030c220, 0x26cc, 0x26cc, 0x64, ...)
	<autogenerated>:79 +0x196
github.com/gonum/lapack/native.Implementation.Dsytrd(0x79, 0x64, 0xc82030c000, 0x2710, 0x2710, 0x64, 0xc82041e700, 0x64, 0x64, 0xc82041ea80, ...)
	/home/daniel/Development/src/github.com/gonum/lapack/native/dsytrd.go:116 +0x5e7
github.com/gonum/lapack/native.(*Implementation).Dsytrd(0x85f6f8, 0x79, 0x64, 0xc82030c000, 0x2710, 0x2710, 0x64, 0xc82041e700, 0x64, 0x64, ...)
	<autogenerated>:65 +0x1a2
github.com/gonum/lapack/testlapack.DsytrdTest(0xc8203a41b0, 0x7fc463a7d128, 0x85f6f8)
	/home/daniel/Development/src/github.com/gonum/lapack/testlapack/dsytrd.go:65 +0x9f1
github.com/gonum/lapack/native.TestDsytrd(0xc8203a41b0)
	/home/daniel/Development/src/github.com/gonum/lapack/native/lapack_test.go:236 +0x76
testing.tRunner(0xc8203a41b0, 0x8431e8)
	/home/daniel/Development/go/src/testing/testing.go:473 +0x98
created by testing.RunTests
	/home/daniel/Development/go/src/testing/testing.go:582 +0x892
```

**Comment by kortschak** [comment]
Apparently due to blas - was at gonum/blas@4a257c6 and updating that to master fixes this - looks like it was fixed by gonum/blas@83fcd69092c941361e712f035c088d9d81ac14e3.
